### PR TITLE
Log warning nlu

### PR DESCRIPTION
**Pull request by chrisbangun**
**Proposed changes**:
- Add the logging mechanism to give a warning if the NLU picks up intent or entities that are not exist in the domain.
- closes https://github.com/RasaHQ/rasa/issues/4119

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by tmbo** [comment]
Thanks a lot for your contribution, we'll give it a review as soon as possible.

File: rasa/core/processor.py [hunk 1]
```diff
@@ -290,6 +290,21 @@ def _log_slots(tracker):
         if slot_values.strip():
             logger.debug("Current slot values: \n{}".format(slot_values))
 
+    def _log_unseen_enitites(self, parse_data):
```
Comment (tabergma) [inline rasa/core/processor.py]: Please add types, e.g. `def _log_unseen_enitites(self, parse_data: Dict[Text, Any]) -> None:`

File: rasa/core/processor.py [hunk 2]
```diff
@@ -290,6 +290,21 @@ def _log_slots(tracker):
...
+        """check if the NLU picks up entities that aren't in the domain.
+        """
+        entities = parse_data["entities"]
+        if not len(entities):
```
Comment (tabergma) [inline rasa/core/processor.py]: No need for the if as the for loop will end immediately if `entities` is empty.

File: rasa/core/processor.py [hunk 3]
```diff
@@ -312,6 +327,16 @@ def _get_action(self, action_name):
                 message.text, parse_data["intent"], parse_data["entities"]
             )
         )
+        # check if we pick up intents that aren't in the domain
```
Comment (tabergma) [inline rasa/core/processor.py]: Please extract this also into a separate function. Same as you did for entities.

File: CHANGELOG.rst [hunk 4]
```diff
@@ -12,6 +12,9 @@ This project adheres to `Semantic Versioning`_ starting with version 1.0.
 
 Added
 -----
+- Added the logging mechanism, to give a warning if the NLU picks up intent or entities
+  that are not exist in the domain. The logic of this warning is implemented in a function called
+  `_log_unseen_enitites` in rasa/core/processor.py.
```
Comment (tabergma) [inline CHANGELOG.rst]: Would rephrase it a bit
```
- log a warning if the ``Interpreter`` picks up an intent or an entity that does not exist in the domain file.
```
No need to point to the function you added. Also no need for the empty line after the entry.

**Comment by chrisbangun** [comment]
> Great work! Left a few comments. Let me know if anything is unclear.
> 
> Also make sure that the tests are passing. Currently, a test is failing here:
> 
> ```
> >       if intent == "" or intent not in self.domain.intents:
> E       AttributeError: 'NoneType' object has no attribute 'intents'
> ```
> 
> You may want to add a null check.

thanks for the input @tabergma . btw, can you have multiple intents in `parse_data`? if it's possible, how is the data format looks like?

thanks

**Comment by tabergma** [comment]
There can be just one `intent` in `parse_data`.

File: rasa/core/processor.py [hunk 5]
```diff
@@ -290,6 +290,30 @@ def _log_slots(tracker):
...
+                "that is not defined in the domain.".format(intent)
+            )
+
+    def _log_unseen_enitites(self, parse_data: Dict[Text, Any]) -> None:
+        """check if the NLU picks up entities that aren't in the domain.
+        """
+        entities = parse_data["entities"]
+        for element in entities:
+            entity = element["entity"]
+            if entity not in self.domain.entities and self.domain:
```
Comment (tabergma) [inline rasa/core/processor.py]:
```suggestion
            if entity and entity not in self.domain.entities and self.domain:
```

Comment (chrisbangun) [reply rasa/core/processor.py]: thanks again @tabergma , have pushed the changes and hopefully it passes the test!

File: rasa/core/processor.py [hunk 6]
```diff
@@ -290,6 +290,30 @@ def _log_slots(tracker):
         if slot_values.strip():
             logger.debug("Current slot values: \n{}".format(slot_values))
 
+    def _log_unseen_intent(self, parse_data: Dict[Text, Any]) -> None:
+        """check if the NLU picks up intent that aren't in the domain.
+        """
+        intent = parse_data["intent"]["name"]
+        if not intent or intent == "":
+            return
+        if intent not in self.domain.intents and self.domain:
```
[the remark this answers is not in the document]
Comment (chrisbangun) [reply rasa/core/processor.py]: thanks again @tabergma , have pushed the changes and hopefully it passes the test!

**Comment by chrisbangun** [comment]
thanks @tabergma and @erohmensing 
It was a nice experience and motivated me to contribute more to Rasa.

Best

Adi

**Comment by erohmensing** [comment]
Sure thing Adi! Thanks for the contribution, looking forward to more from you 😃
